client/orderbook/orderbook.go

```diff
@@ -476,26 +476,49 @@ func (ob *OrderBook) Unbook(note *msgjson.UnbookOrderNote) error {
 }
 
 // BestNOrders returns the best n orders from the provided side.
-// NOTE: This is UNUSED, and test coverage is a near dup of bookside_test.go.
-func (ob *OrderBook) BestNOrders(n int, side uint8) ([]*Order, bool, error) {
+func (ob *OrderBook) BestNOrders(n int, sell bool) ([]*Order, bool, error) {
 	if !ob.isSynced() {
 		return nil, false, fmt.Errorf("order book is unsynced")
 	}
 
 	var orders []*Order
 	var filled bool
-	switch side {
-	case msgjson.BuyOrderNum:
+	if sell {
+		orders, filled = ob.sells.BestNOrders(n)
+	} else {
 		orders, filled = ob.buys.BestNOrders(n)
+	}
 
-	case msgjson.SellOrderNum:
-		orders, filled = ob.sells.BestNOrders(n)
+	return orders, filled, nil
+}
 
-	default:
-		return nil, false, fmt.Errorf("unknown side provided: %d", side)
+// VWAP calculates the volume weighted average price for the specified number
+// of lots.
+func (ob *OrderBook) VWAP(lots, lotSize uint64, sell bool) (avg, extrema uint64, filled bool, err error) {
+	orders, _, err := ob.BestNOrders(int(lots), sell)
+	if err != nil {
+		return 0, 0, false, err
+	}
+
+	remainingLots := lots
+	var weightedSum uint64
+	for _, order := range orders {
+		extrema = order.Rate
+		lotsInOrder := order.Quantity / lotSize
+		if lotsInOrder >= remainingLots {
+			weightedSum += remainingLots * extrema
+			filled = true
+			break
+		}
+		remainingLots -= lotsInOrder
+		weightedSum += lotsInOrder * extrema
 	}
 
-	return orders, filled, nil
+	if !filled {
+		return 0, 0, false, nil
+	}
+
+	return weightedSum / lots, extrema, true, nil
```

Comment:
I remember seeing this code before and commenting that, if the book was sparse, would that allow an attacker to place some orders near the mid gap and then one order way far back, the bot buys them all up because the average looks good, the attacker then doesn't follow through with the extrema trades and just leaves the one at a bad rate. Was this a valid concern? I think my solution was to prevent orders that are too far from the extrema in these calculations.

Reply:
The code here makes sure that each lot included in the arbitrage must be profitable, not just the average:
https://github.com/decred/dcrdex/blob/8f28450c320d7719fce63cece2836696dab437e1/client/mm/mm_simple_arb.go#L218

This would avoid the attack you are talking about right?

Reply:
Oh that's right, the extrema is the worst rate for us, I was wording it as best. Unless both markets are sparse, then Yes this should do about the same thing as I suggested. Would this be even too restrictive? Orders will probably be further apart on dex... If I am understanding correctly.

Reply:
I guess we could do something like buying 1/2 a lots worth on the CEX and a full lots worth on the DEX (since we can't buy any less), but maybe save that for a future iteration.

Unless it's someone who is running the CEX trying to manipulate someone running the arb bot, I don't think it's possible. If somehow they trick us into taking a really bad trade on the DEX, then they would've had to also trick us into taking a really good trade on the CEX. Unless they are running the CEX, they would not be able to back out of the CEX trade like they could on the DEX.